### ISSUE TITLE
rxm providers ignore mode and capabilities bits?

### ISSUE THREAD

**Issue by droccom** [ofiwg/libfabric#3169]
53e5534 Despite I set mode bits to 0 in hints and/or FI_DIRECTED_RECV capability, rxm providers show up from fi_getinfo.

In feature matrix:
https://github.com/ofiwg/libfabric/wiki/Provider-Feature-Matrix-v1.5.0
no rxm providers are shown, thus should I just forget about them until they will be officially supported?

**Comment by shefty** [comment]
I have a request with the maintainer of rxm to update the feature matrix.  That will be updated prior to the 1.5.0 release.  Please go ahead and try using rxm and report any issues that you uncover.  (Consider it officially supported, as the 1.5.0 release is targeted for the next 1-2 weeks.)

For this issue, is rxm returning with mode bits set?  Is FI_DIRECTED_RECV cleared?  I want to make sure the output from fi_getinfo is correct first.  We will continue to work on expanding which features of the API rxm will support even after the 1.5 release.  If you have requirements there, we can use that to help prioritize which new features to implement.

**Comment by a-ilango** [comment]
Rxm provider doesn't require any mode bits by itself but would return any mode bits from core provider except FI_MR_LOCAL and FI_CONTEXT. The provider supports the FI_DIRECTED_RECV capability.

I'm in the process of updating the provider matrix.

**Comment by droccom** [comment]
I did not check the output info (I will do it asap), I just noticed unexpected behaviors:
- context mode in input -> verbs-rxm in output, that crasheh for null context
- directed-recv in input -> verbs/sockets-rxm in output, that ignores src_addr in recv (i post a recv from addr x, i got recv completion upon data from y)

Actually, the directed-recv would be a fundamentalfundamental feature for me, then I would push for it to be prioritized :)

Thank you!

**Comment by shefty** [comment]
I'm having a hard time understanding the actual problems from the above descriptions.

If FI_CONTEXT is specified and returned by the provider, then null context will crash.  That's an api violation.

Does directed receives work over sockets for you?  Which provider combo does it not work?

**Comment by droccom** [comment]
Sorry, it was a typo, i meant context *not* set in input.

The directed-recv works with sockets, it does *not* work with sockets-rxm, then I thought it should not be showed in output.

**Comment by shefty** [comment]
Can you provide a backtrace of the crash that you see when running over verbs;ofi-rxm?

Are you able to test directed_recv with verbs;ofi-rxm?  Or does the crash prevent this from working?

**Comment by a-ilango** [comment]
@droccom 
I've updated the provider feature matrix. If you have a reproducer for the directed recv issue please share it.

**Comment by droccom** [comment]
Thank you,
I will try to be more precise tommorow (Italian working hours)

**Comment by droccom** [comment]
I built the following matrix with the behaviors I observed:
https://docs.google.com/a/unito.it/spreadsheets/d/1WFoXCo4at93JJwNyodlajmJdD3OKYrnlDEPgdJBZfUU/edit?usp=sharing

I considered two applications:
- fabspikes: various configurations of pingpong-like code
- ostream: a stream processor that relies on FI_DIRECTED_RECV for keeping the order among stream items

The matrix is self-explanatory, except for "slow" and "fast" fabspikes executions: in slow executions, I observe a substantial initial latency (about 10 secs) if the server is launched with some delay after the client, whereas I did not observe this in fast executions.

From the matrix, I can conclude that:
- most of the issues I mentioned in the past comments have been addressed
- the story about FI_CONTEXT was my fault, I probably made some confusion at some point
- sockets and sockets-rxm seem to have been fused in a way that dropped some performance gain (i.e. row 14 vs 5)
- [OFF-TOPIC] the verbs provider gets stuck if a receiver posts a recv for which a corresponding send has not been posted (missing some mechanisms of connection retrying?)

I hoe this can help, thank you for your promptness.
